### ISSUE TITLE
Stringparams not being passed?

### ISSUE THREAD

**Issue by StevenClontz** [PreTeXtBook/pretext-cli#566]
I use the setting `<stringparam key="debug.project.number" value="the-way-it-should-be"/>` at https://github.com/TeamBasedInquiryLearning/linear-algebra/blob/main/project.ptx to get project numbers aligned with their siblings. But my build of the 2023 edition and the in-progress 2024 editions have the default asynchronous numbering: [Activity 1.1.1](https://teambasedinquirylearning.github.io/linear-algebra/2023/LE1.html#activity-1)

I'll do some digging to find out when exactly this was affected.

**Comment by StevenClontz** [comment]
Deployment on Aug 12 didn't use the right numbering: https://github.com/TeamBasedInquiryLearning/linear-algebra/blob/ad639c18fd5ce0f997ebda6febb4fc04b0394bf7/2023/AT1.html

Deployment on Jul 21 was correct: https://github.com/TeamBasedInquiryLearning/linear-algebra/blob/6309da040241aa186e8b3521b87701d372f655d0/2023/AT1.html

**Comment by StevenClontz** [comment]
I updated my project.ptx on Aug 14, *after* the numbering issue appeared: https://github.com/TeamBasedInquiryLearning/linear-algebra/commit/aa53214955620a2d5c9995cdff10fc7a93f519a5#diff-1552d8b163a63a46dd5803e3b07b9bc9d300bd7c5ab0a90cda94c332227ea040

**Comment by StevenClontz** [comment]
However, I think perhaps I had updated pretext to 2.0.0rcWhatever manually (not updating my requirements.txt, a mistake I think now), did some experimental builds (which was allowed as nothing tracked by Git was changed), and deployed them. So this seems related to 2.0 changes

**Comment by StevenClontz** [comment]
My guess is that the bug is here: https://github.com/PreTeXtBook/pretext-cli/blob/main/pretext/project/__init__.py#L105

Namely, each stringparam is a `<stringparam/>` element, but we're looking for a `<stringparams/>` element mistakenly.

**Comment by oscarlevin** [comment]
Stringparams must have been working at some point, when @bjones1 implemented the runestone thing, right?  Perhaps there is an issue with attempt to not mutate stringparam dictionaries that happened recently?

**Comment by StevenClontz** [comment]
Possibly. That was Aug 12. https://github.com/PreTeXtBook/pretext-cli/pull/550

**Comment by StevenClontz** [comment]
I also was responsible for this in core on Aug 13 which could be an issue if core is broken: https://github.com/PreTeXtBook/pretext/commit/3f32f9559f4fe8e0b0eb61286cc30576a6d5d4f0

**Comment by StevenClontz** [comment]
I may take a stab later this evening but need to step away. Provisioning a devcontainer so I can at least write a test for this at some point.

**Comment by oscarlevin** [comment]
Okay, so I think I understand what is happening now.  The new way to specify stringparams in `project.ptx` is as follows:

```xml
<target name="web" format="html">
  <stringparams debug.project.number="the-way-it-should-be" html.css.extra="external/custom-styles.css"/>
</target>
```

That is, the `stringparams` element can have attributes, each the name of a stringparam with values for the value of that stringparam.

I suspect this is broken if you have a legacy `project.ptx`, so I will try to fix that ASAP.

@StevenClontz, if you have a good reason that we should keep the old method, then speak up.  I don't know how to do that, but I suspect I'm about to learn to fix the legacy project.

**Comment by StevenClontz** [comment]
Oh I remember some relevant discussion now. Are periods allowed in attribute names?

**Comment by oscarlevin** [comment]
Apparently they are.  It seems to work, at least for v2 project.ptx files

**Comment by StevenClontz** [comment]
Apparently, yes they are. 

> alphabetical characters, decimal digits, underscores, periods (.), and hyphens (-)

@rbeezer is there any reason a stringparam would have a key using other characters?

**Comment by oscarlevin** [comment]
So maybe this is already set up to work for legacy projects.  Take a look at https://github.com/PreTeXtBook/pretext-cli/blob/a4e97b49c278628e5307f65c1320d0a2c8e11b21/pretext/project/xml.py#L45.  I don't understand what is gong on right now, since it doesn't seem to work when I test an old project, but we can't be that far off.

**Comment by rbeezer** [comment]
I am pretty sure I was religious about using periods as separators.  (And then built internal variables with hyphens, so I didn't confuse myself.)

The active `debug.*` series starts at about line 400 of `xsl/pretext-common.xsl`.

> could be an issue if core is broken

Huh???  ;-)

**Comment by bjones1** [comment]
I'll be happy to take a look at this later today. I'm finishing up the first weeh-ish of classes, so it's been crazy.

**Comment by StevenClontz** [comment]
Confirming that `<stringparams debug.project.number="the-way-it-should-be"/>` works as intended.

**Comment by StevenClontz** [comment]
And https://github.com/PreTeXtBook/pretext-cli/blob/6d13ca884e8de9fff6013a3b1050dbce86bbaee3/tests/examples/projects/project_refactor/elaborate/project.ptx#L29 is using stringparams correctly.

**Comment by StevenClontz** [comment]
> I don't understand what is gong on right now, since it doesn't seem to work when I test an old project, but we can't be that far off.

So I think 2.0 stringparams are working correctly, once I use the correct syntax. But I cannot confirm/deny if stringparams in 1.x project.ptx files are broken in 2.0

**Comment by bjones1** [comment]
They look broken to me. Working on a PR.

**Comment by oscarlevin** [comment]
Check out #571 .  I think that fixes it.

**Comment by bjones1** [comment]
Oops, I didn't see this! I'm not sure if that fixes it -- have you tested it? I'll go ahead and make PR; pick what makes sense.

**Comment by bjones1** [comment]
#573

**Comment by oscarlevin** [comment]
Fixed in #574.
